src/downloadAzureProject/setupProjectFolder.ts

```diff
@@ -0,0 +1,115 @@
+import { requestUtils } from '../utils/requestUtils';
+
+export async function setupProjectFolder(uri: vscode.Uri, vsCodeFilePathUri: vscode.Uri): Promise<void> {
+    const parsedQuery: querystring.ParsedUrlQuery = querystring.parse(uri.query);
+    const resourceId: string = getRequiredQueryParameter(parsedQuery, 'resourceId');
+    const devContainerName: string = getRequiredQueryParameter(parsedQuery, 'devcontainer');
+    const language: string = getRequiredQueryParameter(parsedQuery, 'language');
+    const downloadAppContent: string = getRequiredQueryParameter(parsedQuery, 'downloadAppContent');
+
+    ext.context.globalState.update(GlobalStates.projectLanguage, getProjectLanguageForLanguage(language));
+    const toBeDeletedFolderPathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, 'temp');
+
+    try {
+        const functionAppName: string = getNameFromId(resourceId);
+        const downloadFilePath: string = vscode.Uri.joinPath(toBeDeletedFolderPathUri, `${functionAppName}.zip`).fsPath;
+
+        vscode.window.showInformationMessage(localize('settingUpFunctionAppLocalProjInfoMessage', `Setting up project for function app '${functionAppName}' with language '${language}'.`));
+
+        if (downloadAppContent === 'true') {
+            // NOTE: We don't want to download app content for compiled languages.
+            await callWithTelemetryAndErrorHandling('azureFunctions.getFunctionAppMasterKeyAndDownloadContent', async (actionContext: IActionContext) => {
+                const slotTreeItem: SlotTreeItemBase | undefined = await ext.tree.findTreeItem(resourceId, { ...actionContext, loadAll: true });
+                const hostKeys: WebSiteManagementModels.HostKeys | undefined = await slotTreeItem?.client.listHostKeys();
+                const defaultHostName: string | undefined = slotTreeItem?.client.defaultHostName;
+
+                if (!!hostKeys && hostKeys.masterKey && defaultHostName) {
+                    const requestOptions: RequestPrepareOptions = {
+                        method: 'GET',
+                        url: `https://${defaultHostName}/admin/functions/download?includeCsproj=true&includeAppSettings=true`,
+                        headers: { 'x-functions-key':  hostKeys.masterKey }
+                    };
+                    await requestUtils.downloadFile(requestOptions, downloadFilePath);
+                }
+            });
+        }
+
+        const projectFilePathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, `${functionAppName}`);
+        const projectFilePath: string = projectFilePathUri.fsPath;
+        const devContainerFolderPathUri: vscode.Uri = vscode.Uri.joinPath(projectFilePathUri, '.devcontainer');
+        ext.context.globalState.update(GlobalStates.projectFilePath, projectFilePathUri.fsPath);
+        await callWithTelemetryAndErrorHandling('azureFunctions.extractContentAndDownloadDevContainerFiles', async (_actionContext: IActionContext) => {
```

Comment:
> We want to use devContainers for this feature so that customers don't need to install each language specific libraries on their machine.

I'm just skeptical people want Docker on their machine, but I'm pretty confident a JavaScript developer would want Node.js.

> Also, this feature is a two step process and the first step would be an msi which will install all the requirements like Docker, etc on the customer's machine before they hit this flow. 

Does the msi already exist? What about mac/linux? 

> This was mentioned in our initial demo meeting with Matt and you.

Fair enough, but a quick demo meeting does not mean an experience has been finalized.

Reply:
We will be working on the msi on our end and will not enable this feature from our end till we have that ready. We will also have a corresponding file needed for mac/linux as this is our complete experience that we need to ship.

Reply:
As discussed in the meeting, we are not going to enable this feature till we have the msi/pkg piece in.

Reply:
@ejizba Could you provide pointers on how to add feature flag in our handleUri entry point? As that seems like the last thing remaining to get this in

Reply:
It's just a VS Code setting. Define it in package.json:

https://github.com/microsoft/vscode-azureappservice/blob/57ba1a52dc44a351ace10a079ad548770ea66b03/package.json#L850

And reference it inside handleUri like this:

https://github.com/microsoft/vscode-azureappservice/blob/b5a0bdfd6fc4d0bef8dcb6c8167318355f7f88a7/src/commands/remoteDebug/getRemoteDebugLanguage.ts#L20

Reply:
Thanks updated!